docs/src/rfc/text/0017-aml-parser.md

```diff
@@ -0,0 +1,331 @@
+The Rust AML implementation derives its format from the [ACPI SDT protocol](https://uefi.org/specs/PI/1.8/V5_ACPI_System_Desc_Table_Protocol.html),
+which includes basic functionality for parsing and patching AML bytecode after ACPI tables are installed.  
+
+## Goals
+
+Provide a comprehensive Rust implementation for AML parsing and patching on the firmware DXE side.
+Secondarily, implement the rest of the ACPI SDT protocol relating to AML functionality.  
+
+## Requirements
+
+1. Redesign existing C firmware AML interaction functionality into a safe, easy-to-use Rust service.
+2. Implement firmware-side AML parsing: traversal and patching of AML bytecode as opcodes and operands.
+3. Use the Rust service (*1.*) to implement the C ACPI SDT protocol.
+
+## Prior Art
+
+The [ACPI SDT protocol](https://uefi.org/specs/PI/1.8/V5_ACPI_System_Desc_Table_Protocol.html)
+is a spec-defined UEFI PI protocol for retrieving and parsing ACPI tables.
+There are many existing implementations, such as [edk2's AcpiTableDxe](https://github.com/tianocore/edk2/blob/edb5331f787519d1abbcf05563c7997453be2ef5/MdeModulePkg/Universal/Acpi/AcpiTableDxe/AmlChild.c#L4).
+
+An (incomplete) implementation for application-side interpretation of AML bytecode exists in the [Rust `acpi` crate](https://github.com/rust-osdev/acpi).
+While the intent of the `acpi` crate is to interpret and execute AML after boot,
+there is significant overlap in the low-level parsing code between the firmware and application side of AML functionality.
+
+## Alternatives + Open Questions
+
+The [Rust `acpi` crate](https://github.com/rust-osdev/acpi)
+already provides some functionality for interpreting AML bytecode.
+However, it is incomplete and provides limited public interfaces;
+it also does not deal with firmware-side protocols or parsing.
+
+This leaves two main paths for the Patina AML implementation:
+
+1. Design and implement a new Rust AML service from the ground up,
+without explicitly utilizing the existing `acpi` crate.
+(`acpi` has MIT license, so it may be possible to borrow some snippets/implementations with proper attribution.)
+   - Pros: Interfaces and implementations can be tailored to Patina needs.
+   - Cons: Repeated work.
+2. Design and implement the Rust AML service
+while using the `acpi` crate as a dependency and parsing through its public interfaces.
+(This may involve contributing to the `acpi` crate to improve its public interfaces.)
+    - Pros: Less repeated code, especially for parsing.
+    - Cons: `acpi` has limited public interfaces, which may constrain the development of the Rust ACPI service.
+It primarily focuses on looking up and executing AML in the application space,
+with less support for actually walking through and modifying the firmware-side AML object tree.
+
+There is ongoing conversation with the owner of the `acpi` crate about
+borrowing certain implementations and modifying the public interfaces
+to be more friendly to the Patina ACPI implementation.

```

Comment:
Can you elaborate on what "more friendly to the Patina ACPI implementation" means in this context?